File: src/controllers/bookingController.js
const db = require('../db');

exports.getBookings = async (req, res) => {
  const { email } = req.query;
  let query = `
    SELECT b.*, r.name as "roomName", r.location as "location", r.emoji as "emoji"
    FROM bookings b
    JOIN rooms r ON b.room_id = r.id
    JOIN users u ON b.user_id = u.id
  `;
  const params = [];
  
  if (email) {
    query += ` WHERE u.email = $1 ORDER BY b.created_at DESC`;
    params.push(email);
  } else {
    query += ` ORDER BY b.created_at DESC`;
  }

  try {
    const result = await db.query(query, params);
    const formatted = result.rows.map(r => ({
      id: r.id,
      ref: r.ref,
      roomId: r.room_id,
      room: r.roomName,
      location: r.location,
      emoji: r.emoji,
      date: r.date,
      time: r.start_time,
      hrs: r.duration,
      total: r.total_cost,
      status: r.payment_status
    }));
    res.json(formatted);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: err.message });
  }
};

exports.createBooking = async (req, res) => {
  const { ref, name, email, roomId, date, time, hrs, total } = req.body;
  
  if(!email || !name) return res.status(400).json({error: "Email and Name required"});

  try {
    let userResult = await db.query("SELECT id FROM users WHERE email = $1", [email]);
    let userId;
    if (userResult.rows.length === 0) {
      const newU = await db.query("INSERT INTO users (name, email) VALUES ($1, $2) RETURNING id", [name, email]);
      userId = newU.rows[0].id;
    } else {
      userId = userResult.rows[0].id;
    }

    const query = `INSERT INTO bookings (ref, room_id, user_id, date, start_time, duration, total_cost) 
                   VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING id`;
    
    const result = await db.query(query, [ref, roomId, userId, date, time, hrs, total]);
    res.json({ success: true, id: result.rows[0].id });
  } catch(err) {
    console.error(err);
    res.status(500).json({ error: err.message });
  }
};

File: controllers/booking_controller.py
import json
from flask import request, jsonify
from db import get_connection

def get_bookings():
    """GET /bookings - Returns all bookings, optionally filtered by ?email="""
    email = request.args.get("email")
    conn = get_connection()
    cur = conn.cursor()
    try:
        query = """
            SELECT b.id, b.ref, b.room_id, r.name AS room_name, r.location,
                   b.date, b.start_time, b.duration, b.total_cost, b.payment_status
            FROM bookings b JOIN rooms r ON b.room_id = r.id JOIN users u ON b.user_id = u.id
            {} ORDER BY b.created_at DESC
        """
        cur.execute(query.format("WHERE u.email = %s" if email else ""), (email,) if email else ())
        res = [
            {"id":r[0], "ref":r[1], "roomId":r[2], "room":r[3], "location":r[4], "date":r[5], 
             "time":r[6], "hrs":r[7], "total":r[8], "status":r[9]} for r in cur.fetchall()
        ]
        return jsonify(res)
    except Exception as e:
        return jsonify({"error": str(e)}), 500
    finally:
        cur.close()
        conn.close()

def create_booking():
    """POST /bookings - Creates a booking."""
    data = request.get_json()
    ref, name, email = data.get("ref"), data.get("name"), data.get("email")
    room_id, date, time = data.get("roomId"), data.get("date"), data.get("time")
    hrs, total = data.get("hrs"), data.get("total")

    if not email or not name:
        return jsonify({"error": "Email and Name required"}), 400

    conn = get_connection()
    cur = conn.cursor()
    try:
        cur.execute("SELECT id FROM users WHERE email = %s", (email,))
        user = cur.fetchone()
        if user is None:
            return jsonify({"error": "User authentication failed. Please sign in."}), 401
            
        cur.execute("INSERT INTO bookings (ref, room_id, user_id, date, start_time, duration, total_cost) VALUES (%s, %s, %s, %s, %s, %s, %s) RETURNING id", (ref, room_id, user[0], date, time, hrs, total))
        new_id = cur.fetchone()[0]
        conn.commit()
        return jsonify({"success": True, "id": new_id})
    except Exception as e:
        conn.rollback()
        return jsonify({"error": str(e)}), 500
    finally:
        cur.close()
        conn.close()

def delete_booking(booking_id):
    """DELETE /bookings/<id> - Cancels a booking by ID."""
    conn = get_connection()
    cur = conn.cursor()
    try:
        cur.execute("DELETE FROM bookings WHERE id = %s RETURNING id", (booking_id,))
        if cur.fetchone() is None:
            return jsonify({"error": "Booking not found"}), 404
        conn.commit()
        return jsonify({"success": True, "message": "Booking deleted successfully"})
    except Exception as e:
        conn.rollback()
        return jsonify({"error": str(e)}), 500
    finally:
        cur.close()
        conn.close()
